README.md
# Jump
Mario-style game

## Run

* <addr> cabal init
* <addr> cabal install
* <addr> cabal build
* <addr> cabal run Jump

## Controls 

* To start press s 
* To move left, right - arrows 
* To jump - space
* To exit - esc

## Game elements

* Black square - player
* Green cycles - monsters 
* Yellow cycles - coins 
* Tall green rectangle - flag
* Grey squares - boxes
* Red cycles - mushrooms


CHANGELOG.md
# Revision history for Jump

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


src/Back.hs
{-# LANGUAGE ExtendedDefaultRules #-}
{-# LANGUAGE PackageImports #-}

module Back where

import Control.Monad
  ( unless,
    when,
  )
import "GLFW-b" Graphics.UI.GLFW as GLFW


withWindow :: Int -> Int -> String -> (GLFW.Window -> IO ()) -> IO ()
withWindow width height title f = do
  GLFW.setErrorCallback $ Just simpleErrorCallback
  r <- GLFW.init
  when r $ do
    m <- GLFW.createWindow width height title Nothing Nothing
    case m of
      (Just win) -> do
        GLFW.makeContextCurrent m
        f win
        GLFW.setErrorCallback $ Just simpleErrorCallback
        GLFW.destroyWindow win
      Nothing -> return ()
    GLFW.terminate
  where
    simpleErrorCallback e s = putStrLn $ unwords [show e, show s]

readInput :: Window -> ((Bool, Bool, Bool) -> IO ()) -> (Bool -> IO ()) -> IO ()
readInput window directionKeySink skeySink = do
  pollEvents
  l <- keyIsPressed window Key'Left
  r <- keyIsPressed window Key'Right
  u <- keyIsPressed window Key'Space
  unless r (GLFW.setTime 0)
  directionKeySink (l, r, u)
  skeySink =<< keyIsPressed window Key'S

keyIsPressed :: Window -> Key -> IO Bool
keyIsPressed win key = isPress `fmap` GLFW.getKey win key

isPress :: KeyState -> Bool
isPress KeyState'Pressed = True
isPress KeyState'Repeating = True
isPress _ = False

src/Main.hs
{-# LANGUAGE ExtendedDefaultRules #-}
{-# LANGUAGE PackageImports #-}

import Back
import Types
import Control.Concurrent (threadDelay)
import Control.Monad
  ( join,
    unless,
  )
import Control.Monad.Fix (fix)
import FRP.Elerea.Simple
import Game
import Graphics.Gloss.Rendering
import "GLFW-b" Graphics.UI.GLFW as GLFW
import qualified Graphics.UI.GLUT as GLUT
import System.Exit (exitSuccess)

main :: IO ()
main = do
  (_, _) <- GLUT.getArgsAndInitialize
  (directionKeyGen, directionKeySink) <- external (False, False, False)
  (sKeyGen, sKeySink) <- external False
  (windowSizeGen, _) <-
    external
      (fromIntegral width, fromIntegral height)
  glossState <- initState
  withWindow width height "Jump" $ \win -> do
    network <- start $ do
      directionKey <- directionKeyGen
      windowSize <- windowSizeGen
      sKey <- sKeyGen
      jump win windowSize directionKey sKey glossState
    fix $ \loop -> do
      readInput win directionKeySink sKeySink
      join network
      threadDelay 25000
      esc <- keyIsPressed win Key'Escape
      unless esc loop
    exitSuccess

src/Types.hs
module Types where

import           Graphics.Gloss
import           Graphics.Gloss.Data.ViewPort

initialPlayer :: Player
initialPlayer = Player (0, 0) (0, 0) playerSize False

initialLevel :: LevelStatus
initialLevel = Level 1

initialLives :: Int
initialLives = 3

width :: Int
width = 800

height :: Int
height = 500

worldWidth :: Float
worldWidth = 15000

worldHeight :: Float
worldHeight = 500

gravity :: Float
gravity = -5

jumpHeight :: Float
jumpHeight = 100

floorWidth :: Float
floorWidth = 150

floorHeight :: Float
floorHeight = 60

floorLevel :: Float
floorLevel = floorHeight - fromIntegral height / 2

boxSize :: Float
boxSize = 20

coinSize :: Float
coinSize = 10

playerSize :: Float
playerSize = 20

monsterSize :: Float
monsterSize = 20

monsterSpeed :: Float
monsterSpeed = 2.5

mushroomSize :: Float
mushroomSize = 20

data MonsterStatus = Move | Dead | Hit 
                     deriving Eq

data WinGame = WinGame | Other

type Type = Either WinGame

newtype GameState = GameState RenderState

data RenderState = RenderState { renderState_player :: Player
                               , renderState_monster :: [Monster]
                               , renderState_boxes :: [Box]
                               , renderState_floor :: [Box]
                               , renderState_coins :: [Coin]
                               , renderState_Flag :: Flag
                               , renderState_ending :: Maybe Ending
                               , renderState_viewport :: ViewPort
                               , renderState_lives :: Int
                               , renderState_score :: Float
                               , renderState_timer :: Float
                               , renderState_mushrooms :: [Mushroom]
                               , renderState_animation :: Maybe Animation
                               , renderState_windowSize :: (Int, Int) }
                 | StartRenderState (Int, Int)

type Pos = Point

data Animation = DeathAnimation Float | NextLevelAnimation LevelStatus Float
                 deriving Show

data LevelSettings = LevelSettings { levelMonsters :: [Monster]
                                   , levelCoins :: [Coin]
                                   , levelBoxes :: [Box]
                                   , levelFloor :: [Box]
                                   , levelFlag :: Flag
                                   , levelMushrooms :: [Mushroom]
                                   }

data Coin = Coin
  { coin_x :: Float
  , coin_y :: Float
  , vis    :: Bool
  }

data Mushroom = Mushroom
  { mushroom_x :: Float
  , mushroom_y :: Float
  , v          :: Bool
  }

data Flag = Flag
  { flag_x :: Float
  , flag_y :: Float
  , w      :: Bool
  }

data Box = Box
  { box_x      :: Float
  , box_y      :: Float
  , box_width  :: Float
  , box_height :: Float
  }
  deriving Show

data Player = Player
  { position   :: Pos
  , timePlayer :: (Float, Float)
  , size       :: Float
  , bonus      :: Bool
  }

data Ending = Win | Lose
              deriving (Show, Eq)

data GameStatus = Start | InGame
                  deriving Show

newtype LevelStatus = Level Int
  deriving Show

data Monster = Monster
  { pos       :: Pos
  , direction :: Direction
  , monsterStep :: Int
  , isAlive   :: Bool
  }
data Direction = GoLeft {steps :: Int} | GoRight {steps :: Int}


src/Levels.hs
-- file describing levels
{-# LANGUAGE ExtendedDefaultRules #-}

module Levels where

import Graphics.Gloss.Data.ViewPort
import Types
import Control.Monad.Except


helpFunction :: Int -> Type LevelSettings
helpFunction n 
  | n == 1 = return level1
  | n == 2 = return level2
  | otherwise = throwError $ WinGame

getLevel :: Type LevelSettings -> LevelSettings
getLevel (Right level) = level
getLevel (Left er) = level3


initialViewport :: ViewPort
initialViewport =
  ViewPort
    { viewPortTranslate = (0, 0),
      viewPortRotate = 0,
      viewPortScale = 4
    }

initialCoin :: (Float, Float) -> Coin
initialCoin (x, y) = Coin x y True

h :: Float
h = - fromIntegral height / 2

initialMonster :: Pos -> Int -> Monster
initialMonster pos monsterStep = Monster pos (GoLeft monsterStep) monsterStep True

initialBox :: (Float, Float) -> Box
initialBox (x, y) = Box x y boxSize boxSize

initialFloor :: (Float, Float) -> Box
initialFloor (x, y) = Box x y floorWidth floorHeight

level1 :: LevelSettings
level1 = LevelSettings monstersLevel1 coinsLevel1 boxesLevel1 floorLevel1 flagLevel1 mushroomsLevel1

mushroomsLevel1 :: [Mushroom]
mushroomsLevel1 = [Mushroom 270 (140 + h) True]

flagLevel1 :: Flag
flagLevel1 = Flag 4400 (70 + h) False

monstersLevel1 :: [Monster]
monstersLevel1 =
  [ initialMonster (450, 70 + h) 350,
    initialMonster (800, 70 + h) 35,
    initialMonster (1000, 70 + h) 55,
    Monster (855, 70 + h) (GoRight 55) 55 True,
    initialMonster (1500, 70 + h) 50,
    initialMonster (2100, 70 + h) 100,
    initialMonster (2200, 70 + h) 90,
    initialMonster (3900, 70 + h) 135,
    Monster (3560, 70 + h) (GoRight 135) 135 True
  ]

boxesLevel1 :: [Box]
boxesLevel1 =
  [initialBox (200, 120 + h)]
    ++ take 5 [initialBox (250 + x * 20, 120 + h) | x <- [1 ..]]
    ++ [initialBox (450 + x * 20, 50 + y * 20 + h) | x <- [1, 2], y <- [1, 2]]
    ++ [initialBox (650 + x * 20, 50 + y * 20 + h) | x <- [1, 2], y <- [1, 2, 3]]
    ++ [initialBox (800 + x * 20, 50 + y * 20 + h) | x <- [1, 2], y <- [1, 2, 3]]
    ++ [initialBox (1000 + x * 20, 50 + y * 20 + h) | x <- [1, 2], y <- [1, 2, 3]]
    ++ take 4 [initialBox (1450 + x * 20, 120 + h) | x <- [1 ..]]
    ++ take 7 [initialBox (1550 + x * 20, 190 + h) | x <- [1 ..]]
    ++ [initialBox (1870, 120 + h)]
    ++ take 4 [initialBox (1790 + x * 20, 190 + h) | x <- [1 ..]]
    ++ [ initialBox (2780, 130 + h),
         initialBox (2870, 130 + h),
         initialBox (3115, 130 + h),
         initialBox (3285, 130 + h),
         initialBox (4130, 230 + h)
       ]
    ++ [initialBox (1970 + x * 20, 120 + h) | x <- [1, 2]]
    ++ [initialBox (2100 + x * 20, 120 + h) | x <- [1, 3, 5]]
    ++ [initialBox (2160, 170 + h)]
    ++ [initialBox (2300, 120 + h)]
    ++ [initialBox (2400 + x * 20, 190 + h) | x <- [1, 2, 3]]
    ++ [initialBox (2570 + x * 20, 120 + h) | x <- [1, 2]]
    ++ [initialBox (2550 + x * 20, 190 + h) | x <- [1, 2, 3, 4]]
    ++ [initialBox (2700 + x * 20, 70 + h) | x <- [1, 2, 3, 4]]
    ++ [initialBox (2720 + x * 20, 90 + h) | x <- [1, 2, 3]]
    ++ [initialBox (2740 + x * 20, 110 + h) | x <- [1, 2]]
    ++ [initialBox (2850 + x * 20, 70 + h) | x <- [1, 2, 3, 4]]
    ++ [initialBox (2850 + x * 20, 70 + h) | x <- [1, 2, 3, 4]]
    ++ [initialBox (2850 + x * 20, 90 + h) | x <- [1, 2, 3]]
    ++ [initialBox (2850 + x * 20, 110 + h) | x <- [1, 2]]
    ++ [initialBox (3035 + x * 20, 70 + h) | x <- [1, 2, 3, 4]]
    ++ [initialBox (3055 + x * 20, 90 + h) | x <- [1, 2, 3]]
    ++ [initialBox (3075 + x * 20, 110 + h) | x <- [1, 2]]
    ++ [initialBox (3265 + x * 20, 70 + h) | x <- [1, 2, 3, 4]]
    ++ [initialBox (3265 + x * 20, 90 + h) | x <- [1, 2, 3]]
    ++ [initialBox (3265 + x * 20, 110 + h) | x <- [1, 2]]
    ++ [initialBox (3500 + x * 20, 50 + y * 20 + h) | x <- [1, 2], y <- [1, 2]]
    ++ [initialBox (3650 + x * 20, 120 + h) | x <- [1 .. 4]]
    ++ [initialBox (3900 + x * 20, 50 + y * 20 + h) | x <- [1, 2], y <- [1, 2]]
    ++ [initialBox (3950 + x * 20, 70 + h) | x <- [1 .. 9]]
    ++ [initialBox (3970 + x * 20, 90 + h) | x <- [1 .. 8]]
    ++ [initialBox (3990 + x * 20, 110 + h) | x <- [1 .. 7]]
    ++ [initialBox (4010 + x * 20, 130 + h) | x <- [1 .. 6]]
    ++ [initialBox (4030 + x * 20, 150 + h) | x <- [1 .. 5]]
    ++ [initialBox (4050 + x * 20, 170 + h) | x <- [1 .. 4]]
    ++ [initialBox (4070 + x * 20, 190 + h) | x <- [1 .. 3]]
    ++ [initialBox (4090 + x * 20, 210 + h) | x <- [1 .. 2]]

coinsLevel1 :: [Coin]
coinsLevel1 =
  [ initialCoin (200, 140 + h),
    initialCoin (310, 230 + h),
    initialCoin (1490, 140 + h),
    initialCoin (1650, 210 + h),
    initialCoin (1870, 140 + h),
    initialCoin (2160, 190 + h),
    initialCoin (3710, 140 + h)
  ]
    ++ [initialCoin (250 + x * 20, 140 + h) | x <- [3, 5]]
    ++ [initialCoin (2100 + x * 20, 140 + h) | x <- [1, 3, 5]]
    ++ [initialCoin (2570 + x * 20, 210 + h) | x <- [1, 2]]

floorLevel1 :: [Box]
floorLevel1 =
  take
    35
    [ initialFloor
        ( (- fromIntegral width) / 2 + x * floorWidth,
          floorHeight / 2 + h
        )
      | x <- [0 .. 10] ++ [12, 13] ++ [15 .. 23] ++ [25 ..]
    ]

level2 :: LevelSettings
level2 = LevelSettings monstersLevel2 coinsLevel2 boxesLevel2 floorLevel2 flagLevel2 mushroomsLevel2

mushroomsLevel2 :: [Mushroom]
mushroomsLevel2 = [Mushroom 1170 (150 + h) True]

flagLevel2 :: Flag
flagLevel2 = Flag 5200 (70 + h) False

monstersLevel2 :: [Monster]
monstersLevel2 =
  [ initialMonster (325, 70 + h) 200,
    Monster (100, 70 + h) (GoRight 93) 93 True,
    initialMonster (630, 70 + h) 25,
    initialMonster (1100, 70 + h) 150,
    Monster (710, 70 + h) (GoRight 200) 200 True,
    initialMonster (1770, 70 + h) 210,
    Monster (1270, 70 + h) (GoRight 200) 200 True,
    initialMonster (2280, 70 + h) 130,
    initialMonster (2260, 70 + h) 130,
    initialMonster (2580, 70 + h) 36,
    initialMonster (4620, 70 + h) 175
  ]

boxesLevel2 :: [Box]
boxesLevel2 =
  [initialBox (200 + x * 20, 120 + h) | x <- [1 .. 5]]
    ++ [ initialBox (350, 70 + h),
         initialBox (600, 150 + h),
         initialBox (840, 140 + h),
         initialBox (900, 140 + h)
       ]
    ++ [initialBox (390, 70 + y * 20 + h) | y <- [0, 1]]
    ++ [initialBox (430, 70 + y * 20 + h) | y <- [0 .. 2]]
    ++ [initialBox (470, 70 + y * 20 + h) | y <- [0 .. 3]]
    ++ [initialBox (510, 70 + y * 20 + h) | y <- [0 .. 3]]
    ++ [initialBox (550, 70 + y * 20 + h) | y <- [0 .. 2]]
    ++ [initialBox (650, 70 + y * 20 + h) | y <- [0 .. 2]]
    ++ [initialBox (690, 70 + y * 20 + h) | y <- [0, 1]]
    ++ [initialBox (800 + x * 20, 120 + h) | x <- [0 .. 2]]
    ++ [initialBox (840 + x * 20, 160 + h) | x <- [0 .. 3]]
    ++ [initialBox (900 + x * 20, 120 + h) | x <- [0 .. 2]]
    ++ [initialBox (1100 + x * 20, 120 + h + 20 * y) | x <- [0 .. 2], y <- [0 .. 5]]
    ++ [initialBox (1160 + x * 20, 90 + h + 20 * y) | x <- [0 .. 2], y <- [0 .. 2]]
    ++ [initialBox (1260 + x * 20, 130 + h) | x <- [0 .. 3]]
    ++ [initialBox (1340 + x * 20, 130 + h + 20 * y) | x <- [0 .. 2], y <- [0 .. 5]]
    ++ [initialBox (1500, 130 + y * 20 + h) | y <- [0 .. 5]]
    ++ [initialBox (1520, 130 + h)]
    ++ [initialBox (1540, 130 + h + y * 20) | y <- [0, 1]]
    ++ [initialBox (1600 + x * 20, 130 + h + 20 * y) | x <- [0 .. 2], y <- [0 .. 5]]
    ++ [initialBox (1720 + x * 20, 130 + h) | x <- [0 .. 2]]
    ++ [initialBox (1840 + x * 20, 160 + h + 20 * y) | x <- [0 .. 5], y <- [0 .. 1]]
    ++ [initialBox (2300 + x * 20, 70 + h + 20 * y) | x <- [0 .. 1], y <- [0 .. 2]]
    ++ [initialBox (2450 + x * 20, 70 + h + 20 * y) | x <- [0 .. 1], y <- [0 .. 3]]
    ++ [initialBox (2600 + x * 20, 70 + h + 20 * y) | x <- [0 .. 1], y <- [0 .. 1]]
    ++ [initialBox (3635 + x * 20, 70 + h) | x <- [0 .. 4]]
    ++ [initialBox (3655 + x * 20, 90 + h) | x <- [0 .. 3]]
    ++ [initialBox (3675 + x * 20, 110 + h) | x <- [0 .. 2]]
    ++ [initialBox (3695 + x * 20, 130 + h) | x <- [0 .. 1]]
    ++ [initialBox (3920 + x * 20, 50 + h) | x <- [0 .. 3]]
    ++ [initialBox (3900 + x * 20, 130 + h) | x <- [0 .. 3]]

coinsLevel2 :: [Coin]
coinsLevel2 =
  [initialCoin (220, 140 + h), initialCoin (600, 170 + h), initialCoin (820, 140 + h)] 
    ++ [initialCoin (840 + x * 20, 180 + h) | x <- [0 .. 3]]
    ++ [initialCoin (1260 + x * 20, 150 + h) | x <- [0 .. 3]]
    ++ [initialCoin (1520, 150 + h)]
    ++ [initialCoin (1840 + x * 20, 200 + h) | x <- [0 .. 5]]
    ++ [initialCoin (3920 + x * 20, 70 + h) | x <- [0 .. 3]]

floorLevel2 :: [Box]
floorLevel2 =
  take
    35
    [ initialFloor
        ( (- fromIntegral width) / 2 + x * floorWidth,
          floorHeight / 2 + h
        )
      | x <- [0 .. 14] ++ [16 .. 21] ++ [23] ++ [25 .. 27] ++ [31 .. 33] ++ [35 ..]
    ]

level3 :: LevelSettings 
level3 = LevelSettings [] [] [] [] flagLevel3 []

flagLevel3 :: Flag
flagLevel3 = Flag 5000 (70 - fromIntegral height / 2) False


src/Render.hs
{-# LANGUAGE ExtendedDefaultRules #-}
{-# LANGUAGE PackageImports #-}

module Render where

import Graphics.Gloss
import Graphics.Gloss.Data.ViewPort
import Graphics.Gloss.Rendering
import "GLFW-b" Graphics.UI.GLFW as GLFW
import Types

-- main function in graphics
renderFrame :: GLFW.Window -> State -> (Float, Float) -> RenderState -> IO ()
renderFrame window glossState _ (RenderState player monsters boxes floors coins flag gameOver viewport lives score timer mushrooms mbAnimation dimensions ) =
  do
    displayPicture dimensions white glossState (viewPortScale viewport) $
      Pictures $
        animation mbAnimation dimensions $
          gameOngoing gameOver $
            gameStats
              lives
              score
              timer
              dimensions
              [ renderPlayer player,
                uncurry translate (viewPortTranslate viewport) $
                  Pictures $
                    map
                      renderMonster
                      (filter isAlive monsters),
                uncurry translate (viewPortTranslate viewport) $
                  Pictures $
                    map
                      renderCoin
                      (filter vis coins),
                uncurry translate (viewPortTranslate viewport) (renderFlag flag),
                uncurry translate (viewPortTranslate viewport) $
                  Pictures $
                    map
                      renderBox
                      boxes,
                uncurry translate (viewPortTranslate viewport) $
                  Pictures $
                    map
                      renderFloor
                       floors,
                uncurry translate (viewPortTranslate viewport) $
                  Pictures $
                    map
                      renderMushroom
                      (filter v mushrooms)
              ]
    swapBuffers window
renderFrame window glossState _ (StartRenderState dimensions) = do
  displayPicture dimensions black glossState 1 $
    Pictures
      [ Color green $ translate (-140) 0 $ scale 0.4 0.4 $ Text "Jump",
        Color green $ translate (-140) (-50) $ scale 0.1 0.1 $ Text "Press s"
      ]
  swapBuffers window

renderMushroom :: Mushroom -> Picture
renderMushroom (Mushroom x y _) = 
  Color red $ translate x y $ circleSolid (mushroomSize / 2)

renderFlag :: Flag -> Picture
renderFlag (Flag xf yf _) =
  Color green $ translate xf yf $ rectangleSolid 10 300

renderMonster :: Monster -> Picture
renderMonster (Monster (x, y) _ _ _) =
  Color green $ translate x y $ circleSolid (monsterSize / 2)

renderCoin :: Coin -> Picture
renderCoin (Coin x y _) = Color yellow $ translate x y $ circleSolid coinSize

renderFloor :: Box -> Picture
renderFloor (Box x y w h) = Color orange $ translate x y $ rectangleSolid w h

renderBox :: Box -> Picture
renderBox (Box x y w h) = Color (greyN 0.5) $ translate x y $ rectangleSolid w h

renderPlayer :: Player -> Picture
renderPlayer (Player (_, ypos) _ pSize _) =
  Color black $ translate 0 ypos $ rectangleSolid pSize pSize

-- finish level
gameOngoing :: Maybe Ending -> [Picture] -> [Picture]
gameOngoing (Just Lose) pics =
  pics ++ [Color black $ translate (-100) 0 $ Scale 0.3 0.3 $ Text "Game Over"]
gameOngoing (Just Win) pics =
  pics ++ [Color black $ translate (-100) 0 $ Scale 0.3 0.3 $ Text "Win, next level"]
gameOngoing Nothing pics = pics

-- store, lives, timer
gameStats :: Int -> Float -> Float -> (Int, Int) -> [Picture] -> [Picture]
gameStats lives score timer (w, h) pics = do
  let fWidth = fromIntegral w
      fHeight = fromIntegral h
  pics
    ++ [ Color black $
           translate (fWidth / 2 - 150) (fHeight / 2 - 50) $
             Scale 0.2 0.2 $
               Text $
                 "Score: " ++ show (round score),
         Color black $
           translate (-100) (fHeight / 2 - 50) $
             Scale 0.2 0.2 $
               Text $
                 "Timer: " ++ show (round timer),
         Color (makeColor 1 1 1 0.5) $
           translate ((- fWidth / 2) + 80) (fHeight / 2 - 40) $
             rectangleSolid 250 40,
         Color black $
           translate ((- fWidth) / 2 + 20) (fHeight / 2 - 50) $
             Scale 0.2 0.2 $
               Text $
                 "Lives: " ++ show lives
       ]


animation :: Maybe Animation -> (Int, Int) -> [Picture] -> [Picture]
animation Nothing _ pics = pics
animation (Just (DeathAnimation _)) _ pics = pics
animation (Just (NextLevelAnimation l n)) (w, h) pics =
  pics
    ++ [ Color (animationColor n) $
           rectangleSolid (fromIntegral w) (fromIntegral h),
         Color white $ translate (-100) 0 $ scale 0.3 0.3 $ Text $ show l
       ]
  where
    animationColor i
      | n > 25 = makeColor 0 0 0 (0.04 * (50 - i))
      | otherwise = makeColor 0 0 0 (0.04 * i)